Fixes #10361 - Introduce QoSHandler

sbordet opened the pull request:
* Introduced QoSHandler.
* Updated documentation, and removed old documentation.

sbordet committed:
Fixes #10361 - Introduce QoSHandler

* Introduced QoSHandler.
* Updated documentation, and removed old documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,369 @@
+ * Priorities are determined via {@link #getPriority(Request)}, that
+ * returns values between {@code 0} (lowest) and the value set via
+ * {@link #setMaxPriority(int)} (highest).</p>
+ * <p>This filter is ideal to avoid contending on slow/limited
```
```suggestion
 * <p>This handler is ideal to avoid contending on slow/limited
```

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    /**
+     * <p>Sets the max duration of time a request may stay suspended.</p>
+     * <p>{@link Duration#ZERO} means that the request may stay suspended forever.</p>
+     *
+     * @param maxSuspend the max duration of time a request may stay suspended
```
need to say what happens when the duration expires

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,369 @@
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} forwarding {}", this, request);
+            request.addHttpStreamWrapper(stream -> new Resumer(stream, request));
```
At the very least we need to document that "concurrent request" means not just during handling, but all the way through to the completion of the callback.

However, I'm not convinced this will always be the best solution if there are asynchronous handlers involved (and any webapp that serves static content might have an async handler).

Would it be too much to have the option to resume either on completion or on return from handling?

sbordet replied:
I'd stay minimal for this iteration.

Async handlers may do all kind of things, for example consume CPU in a tight loop while processing large amounts of data, so I am not sure I want to allow other requests in, even if the handler returned.

I chose the most conservative approach for now, we can relax later if there is request for it.

gregw replied:
Ok, but document please

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,369 @@
+
+    private void resume()
+    {
+        semaphore.release();
```
since it is this thread that is going to do the `processAndComplete` do we really need to release the semaphore and then look for an entry?  isn't there a risk that a newly arrived job will steal the semaphore even if it has a lower priority, before we acquire it again.

Can't we just keep the semaphore and go on to process the highest priority entry.  We should only release if we don't find such an entry.

sbordet replied:
Good point, I'll do that.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,369 @@
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} timeout {}", QoSHandler.this, request);
+                Response.writeError(request, response, callback, HttpStatus.SERVICE_UNAVAILABLE_503);
```
We need to protect error page handling from this mechanism so that they don't need a semaphore

sbordet replied:
Not sure I understand what you meant here.

gregw replied:
If you can't get a semaphore, then you can writeError, which may be an error page dispatch that will try to get the semaphore again which is not available.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,369 @@
+        @Override
+        protected Iterator<Entry> iterator()
+        {
+            return queues.stream()
+                .flatMap(Queue::stream)
+                .iterator();
```
Can we avoid streams?  I know it will be a bit more code, but this iterator is called frequently

sbordet replied:
By default this iterator is never called.
When there is a max suspend duration, it is called roughly every max suspend duration, which is typically not that often.
The correspondent non-stream code would be quite complicated, as it would need to return an iterator that iterates over multiple other iterators, which would be a lot more than 3 lines of code.

gregw replied:
Ok, but can you say that in a comment so we don't forget

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,369 @@
+            return expireNanoTime;
+        }
+
+        private void expire()
```
Can this be delegated to a protected method on the handler so expiration actions can be extended?   Or is there not much they can do other than write and error?  I guess they could do something like dispatch handling to a low priority thread pool?

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,369 @@
+     * @param maxPriority the max priority of suspended requests
+     * @see #getPriority(Request)
+     */
+    public void setMaxPriority(int maxPriority)
```
This feels like exposing an implementation detail that we use a fixed size array list.
We could use a sorted data structure that would not need this method.   Hmm but might be worth the efficiency of a final array list of queues? Maybe better as a constructor argument?

sbordet replied:
Removed `maxPriority`, now using a `ConcurrentHashMap` and a side `Set` in reverse iteration order.
Using `ConcurrentSkipListMap` in reverse iteration order would have been more compact, but unfortunately its `compute()` is not atomic -- a surprise.

gregw reviewed (changes requested):
A few thoughts, but mostly it looks OK.

sbordet committed:
Updates after review.

* Ditched Semaphore in favor of AtomicBiInteger.
This also fixed a race condition described in code comments.
* Now using dynamic max priority.
The current implementation retains priorities and queue as they are created, but hopefully the priority range is constrained.
* Better algorithm for resuming: rather than releasing a permit and try to acquire it again, keep the permit and resume a request.
This also removes the issue that a request could have been suspended again, therefore remaining suspended much more than maxSuspend.
* Now using an IteratingCallback to resume requests to avoid StackOverflowError.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed missing return statement in while+cas loop.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw reviewed (changes requested):
Mostly just a few more comments

gregw commented:
II also think this will work well with a `ConditionalHandlerWrapper`.  I'll prepare a PR today.

gregw commented:
@sbordet can you look at draft #10492 and think about how that might apply to this QoSHandler?

sbordet committed:
Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+
+            // Here there are no suspended requests, so increment the permits.
+            // See correspondent code in handle().
+            long encoded = state.get();
```
The thread-safety logic looks right, but the amount of complex mechanisms in this class (permits&suspension atomic bi-counter, IteratingCallback, queues for priorities) makes me worried that there is too much complex logic in this class.

gregw replied:
Ultimately I think the permits/suspend/resume logic should be factored out to a utility class that is also used by `DoSHandler` and `ThreadLimitHandler`.

If there is time, I'm happy to see that in this release cycle, but I'm also OK for that to be done in the next cycle as well (just let's not forget to do so).

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+        // so that already suspended requests will not starve.
+        while (true)
+        {
+            for (Integer priority : priorities)
```
This logic looks like a scheduler. It may be worth extracting that into a utility class.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+                while (true)
+                {
+                    int suspended = state.getLo();
+                    if (!state.compareAndSetLo(suspended, suspended - 1))
```
More state-handling logic...

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+    /**
+     * @return the max duration of time a request may stay suspended
+     */
+    public Duration getMaxSuspend()
```
Why is this a `Duration` instead of a simple `long`?

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+            this.callback = callback;
+            this.priority = priority;
+            Duration maxSuspend = getMaxSuspend();
+            this.expireNanoTime = maxSuspend.isZero() ? Long.MAX_VALUE : NanoTime.now() + maxSuspend.toNanos();
```
What if `NanoTime.now() + maxSuspend.toNanos()` == `Long.MAX_VALUE`?

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+
+/**
+ * <p>A quality of service {@link Handler} that limits the number
+ * of concurrent requests, to provide good end-user experience in
```
_to provide  ~~good~~ **predictable** end-user experience_

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+     * @param request the suspended request to compute the priority for
+     * @return the priority of the given suspended request, a value {@code >= 0}
+     */
+    protected int getPriority(Request request)
```
There are no checks that `getPriority` won't return negative values. Also, why are those forbidden?

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+public class QoSHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(QoSHandler.class);
+    private static final String EXPIRED_ATTRIBUTE = QoSHandler.class.getName() + ".expired";
```
Rename to `EXPIRED_ATTRIBUTE_NAME`?

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+    }
+
+    @ManagedAttribute("The number of suspended requests")
+    public long getSuspendedRequests()
```
Rename to `getSuspendedRequestCount()`?

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+     * @return the max number of concurrent requests
+     */
+    @ManagedAttribute(value = "The maximum number of concurrent requests", readonly = true)
+    public int getMaxRequests()
```
Rename to `getMaxConcurrentRequestCount()`, or `getMaxConcurrency()`?

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+     * @param status the failure status code
+     * @param failure the failure
+     */
+    protected void fail(Request request, Response response, Callback callback, int status, Throwable failure)
```
Since this is API, maybe rename to a more explicit name like `failSuspendedRequest()`?

lorban commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/QoSHandlerTest.java:
```diff
@@ -0,0 +1,320 @@
+
+    private void start(QoSHandler qosHandler) throws Exception
+    {
+        server = new Server();
```
I wonder if it would be worth adding buffer leak tracking logic here.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,465 @@
+    private int maxRequests;
+    private Duration maxSuspend = Duration.ZERO;
+
+    public QoSHandler()
```
Please add a constructor that takes a `Handler` to pass it to `super`. This helps chaining handlers.

lorban commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/QoSHandlerTest.java:
```diff
@@ -0,0 +1,320 @@
+            """.formatted(maxRequests));
+        endPoints.add(endPoint);
+
+        assertEquals(maxRequests, callbacks.size());
```
Shouldn't this assertion also use awaitility, otherwise it may be flaky?

sbordet committed:
Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Deprecated QoSFilters in ee10 and ee9.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,486 @@
+
+        while (true)
+        {
+            long encoded = state.get();
+            int permits = AtomicBiInteger.getHi(encoded);
+            if (permits > 0)
+            {
+                if (!state.compareAndSet(encoded, permits - 1, 0))
+                    continue;
+                return forward(request, response, callback);
+            }
```
As per our conversation last night I'm still a bit worried by this code as you are assuming that suspended == 0, but without checking it.  If by some chance it was not zero in the encoded, you overwrite it to zero here!
So I'm definitely -1 on the current code as it is fragile, but we don't need to go the full way to what I described on the hangout.

But at the very least this needs to be:

```suggestion
            long encoded = state.get();
            int permits = AtomicBiInteger.getHi(encoded);
            int suspended = AtomicBiInteger.getLo(encoded);
            if (permits > 0 && suspended == 0)
            {
                if (!state.compareAndSet(encoded, permits - 1, 0))
                    continue;
                return forward(request, response, callback);
            }
```

If there are other suspended requests, then we will suspend this request, which is OK... but does not handle the case of excess permits still being available

The alternative I prefer is:

```suggestion
            long encoded = state.get();
            int permits = AtomicBiInteger.getHi(encoded);
            int suspended = AtomicBiInteger.getLo(encoded);
            if (permits > suspended)
            {
                if (!state.compareAndSet(encoded, permits - 1, suspended))
                    continue;
                return forward(request, response, callback);
            }
```

sbordet replied:
@gregw the CAS will _not_ overwrite the suspended value with `0`.
The invariant we have is that only when `permits` goes to 0, then `suspended` can be incremented, and vice-versa.
So your first suggestion is ok, the second (while working) is not because the invariant is not clear anymore: the 2 values cannot be updated independently, you can only update one if the other is exactly zero.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,486 @@
+                    // Below, the suspended request count is only incremented if there are no permits.
+                    // See correspondent code in resume() and expire().
+                    int suspended = AtomicBiInteger.getLo(encoded);
+                    if (!state.compareAndSet(encoded, 0, suspended + 1))
```
This is OK with the current fragile code.  But with either of the alternatives I suggest above, this needs to be:
```suggestion
                    if (!state.compareAndSet(encoded, permits, suspended + 1))
```

I know you like the `0`s, but I actually prefer seeing `state.compareAndSet(encoded, function(permits), function(suspended)` as hard coded values like `0` are just assuming the value in encoded and are thus fragile.

sbordet replied:
It's not fragile, is it explicitating the invariant.

gregw replied:
@sbordet It is not making invariant explicit.  Instead it is assuming the invariant and hence it is fragile.  If the invariant ever changes, then this code will break.  To make the invariant explicit it needs to extract the permits field and check that it is zero.   Currently is just overwriting whatever the value is to zero because it assumes it is zero.

Maybe a single AtomicInteger would be better, with positive values representing permits and negative values representing the suspended count.   Then the invariant could never be broken as you could not have both permits and suspended at the same time.   If your state variables allow illegal states, then you need to check that those states are not entered rather than assume they are not.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,486 @@
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} no suspended requests to resume", this);
+                return;
+            }
```
again this hard coded zero itches me when you have not actually checked that permits are zero.  You are only inferring it because you didn't find a suspended request and thus assume encoded will change if the suspended requests change.   Far better to be explicit:
```suggestion
            int permits = AtomicBiInteger.getHi(encoded);
            int suspended = AtomicBiInteger.getLo(encoded);
            if (suspended > 0)
                continue; // spin and look for a suspended request again
            if (state.compareAndSet(encoded, permits + 1, suspended))
            {
                if (LOG.isDebugEnabled())
                    LOG.debug("{} no suspended requests to resume", this);
                return;
            }
```

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,486 @@
+        }
+    }
+
+    private boolean tryResumeSuspended()
```
```suggestion
    private boolean resumeSuspended()
```

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,486 @@
+
+    private class Forwarder extends IteratingCallback
+    {
+        private final Queue<Entry> queue = new ConcurrentLinkedQueue<>();
+        private Entry entry;
+
+        private void offer(Entry entry)
+        {
+            queue.offer(entry);
+            iterate();
+        }
+
+        @Override
+        protected Action process()
+        {
+            entry = queue.poll();
+            if (entry == null)
+                return Action.IDLE;
+
+            entry.resume(this);
+            return Action.SCHEDULED;
+        }
```
This should not need to be a queue. There should only be either 0 or 1 resumed requests.
```suggestion
        private final AtomicReference<Entry> resumed = new AtomicReference<>();
        private Entry entry;

        private void offer(Entry entry)
        {
            if (!resumed.compareAndSet(null, entry))
                throw new IllegalStateException();
            iterate();
        }

        @Override
        protected Action process()
        {
            entry = resumed.getAndSet(null);
            if (entry == null)
                return Action.IDLE;

            entry.resume(this);
            return Action.SCHEDULED;
        }
```

sbordet replied:
It must be a queue.
If 10 requests are processed by 10 different threads, when they are complete you have 10 threads calling `resume()`, which will find 10 meanwhile suspended requests that are offered to the `Forwarder`.

Perhaps we may be smarter and if the thread is different we don't use the `Forwarder`? 
Otherwise the requests that may be concurrent are serialized.

gregw replied:
Hmmm now it feels more like a ThreadLocal.
If we have 10 different requests all with their callbacks being completed at the same time, then for each request there are two options:
1. the thread that called `Handler.handle(...)` is the one completing, in that case we only need to take one job from the suspended queue and remember it until we return from handle, when we can call the next `handle`.  Of course this assumes that after completing the callback that the thread returns immediately and does not go blocking on databases trying to commit some result.
2. the thread completing is some other thread not in the scope, in which case we could just make it execute any resumed job itself... or is that a breach of the Callback contract as it may take some time?  Perhaps in that case we need to execute a new thread?

Eitherway, I don't see that we should be taking entries off the suspended queue only to add them to an iteration queue.  Double contention!

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,486 @@
+        Response.writeError(request, response, callback, status, null, failure);
+    }
+
+    private boolean forward(Request request, Response response, Callback callback) throws Exception
```
I still don't like the name.  Anybody that works with Servlet code a lot will just keep thinking of a `RequestDispatcher.forward` and get confused. 

 How about:

```suggestion
    private boolean handleWithPermit(Request request, Response response, Callback callback) throws Exception
```

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,486 @@
+        }
+    }
+
+    private class Forwarder extends IteratingCallback
```
```suggestion
    private class ResumedProcessor extends IteratingCallback
```

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Optimization for the async case.
Added TODO to discuss what would be the best solution.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,477 @@
+    private class Resumer extends HttpStream.Wrapper
+    {
+        private final Request request;
+        private final Thread thread;
```
```suggestion
        private final Thread handlingThread;
```

gregw reviewed (changes requested):
This is so much more complex that how complex we thought it was!

gregw commented:
@sbordet I'm wondering if we should fall back to always executing a resumed request.   Then add in some optimisations to reuse threads only when we are sure it is all correct and working fine?

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,432 @@
+    {
+        // See correspondent state machine logic in handle() and expire().
+        int permits = state.getAndIncrement();
+        if (permits >= 0)
```
Just for the sake of clarity, I think doing the latter would be more readable:
```java
int permits = state.incrementAndGet();
if (permits > 0)
```

lorban reviewed (approved):
LGTM, except for one nit.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,432 @@
+ * always be able to access the web application.</p>
+ */
+@ManagedObject
+public class QoSHandler extends Handler.Wrapper
```
This could extend ConditionalHandler.... just saying :)

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,432 @@
+        int permits = state.getAndDecrement();
+        if (permits > 0)
+        {
+            return forward(request, response, callback);
```
I think we need to make it very clear that we have taken a permit at line 188, not just checked how many there are.  Either comment that or just rename:
```suggestion
            return handleWithPermit(request, response, callback);
```

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,432 @@
+        }
+    }
+
+    private static void notAvailable(Response response, Callback callback)
```
```suggestion
    protected static void notAvailable(Request request, Response response, Callback callback)
```

gregw replied:
@sbordet can you do this one?

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,432 @@
+        for (Integer priority : priorities)
+        {
+            Queue<Entry> queue = queues.get(priority);
+            Entry entry = queue.poll();
```
There is a tiny window for a NPE here as a new priority is added to the `priorities` just before it creates the new entry in `queues`, thus `queue` might end up being null here.
If it is, I think it is safe to return false, as the caller will just spin and call again

sbordet replied:
@gregw we also might see the queue but not the entry, so I modified the `offer()` side too.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java:
```diff
@@ -0,0 +1,432 @@
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} succeeded {}", QoSHandler.this, request);
+            resume();
```
I'm just thinking out-loud here for future optimizations.... If the thread here was the same thread that called `handle` and that then called `handleWithPermit`, then we could delay calling `resume()` here and somehow set state so that when we return to handle, it will then call a new `resumeWithPermit` method.

I think that might work for a very common case.  But I'm not sure how much it would gain.

gregw reviewed (changes requested):
I think I found a tiny weeny race that might result in a NPE... plus a few niggles.

sbordet committed:
Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw reviewed (approved):
Looks good. But can you make not available protected so it can be extended.

sbordet commented:
@gregw, I kept `notAvailable()` private because it is a very special case and if overridden doing `Response.writeError()` ends up in an infinite loop.

Timeouts are delivered to `failSuspended()` anyway, so users have the chance to override and customize if they want to.
